src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNIndexCache.java

```diff
@@ -175,18 +178,57 @@ public void setCacheCapacityReached(Boolean value) {
      * Loads hnsw index to memory. Registers the location of the serialized graph with ResourceWatcher.
      *
      * @param indexPathUrl path for serialized hnsw graph
-     * @param algoParams hnsw algoparams
+     * @param algoParams hnsw algorithm parameters
      * @return KNNIndex holding the heap pointer of the loaded graph
      * @throws Exception Exception could occur when registering the index path
      * to Resource watcher or if the JNI call throws
      */
-    public KNNIndex loadIndex(String indexPathUrl, final String[] algoParams) throws Exception {
+    public KNNIndexCacheEntry loadIndex(String indexPathUrl, final String[] algoParams) throws Exception {
         if(Strings.isNullOrEmpty(indexPathUrl))
             throw new IllegalStateException("indexPath is null while performing load index");
         logger.debug("Loading index on cache miss .. {}", indexPathUrl);
         Path indexPath = Paths.get(indexPathUrl);
-        knnIndexFileListener.register(indexPath);
-        return KNNIndex.loadIndex(indexPathUrl, algoParams);
+        FileWatcher fileWatcher = new FileWatcher(indexPath);
+        fileWatcher.addListener(KNN_INDEX_FILE_DELETED_LISTENER);
+
+        // Calling init() on the FileWatcher will bootstrap initial state that indicates whether or not the file
+        // is present. If it is not present at time of init(), then KNNIndex.loadIndex will fail and we won't cache
+        // the entry
+        fileWatcher.init();
+
+        final KNNIndex knnIndex = KNNIndex.loadIndex(indexPathUrl, algoParams);
+
+        // TODO verify that this is safe - ideally we'd explicitly ensure that the FileWatcher is only checked
+        // after the guava cache has finished loading the key to avoid a race condition where the watcher
+        // causes us to invalidate an entry before the key has been fully loaded.
+        final WatcherHandle<FileWatcher> watcherHandle = resourceWatcherService.add(fileWatcher);
```

Comment:
I think guava cache takes the lock when loading the entry for the key. So when invalidate happens, it should happen after the graph is loaded? 
 https://github.com/google/guava/blob/master/guava/src/com/google/common/cache/Cache.java#L91

Reply:
See the following gist. Guava invalidate(key) does not block but is instead a no-op when executed if guava hasn't fully loaded the key into the cache: https://gist.github.com/jschmitz28/c8ff2d3b5856f625cf1f5a98095b7a51

Reply:
Got it. You mean invalidating entry would not interrupt loading graph but no-op would make us miss ```gc()``` on the graph.  Agree this minimizes the risk of race condition assuming the index deletion happens during search.